### PR TITLE
SFR-1232: Adding e2e tests for Clever login

### PR DESCRIPTION
**Pull request by samanthaandrews**
- Adds e2e tests for logging in with Clever
- I'll update this PR to include a couple integration tests once I pair with the team to understand the Clever login flow a bit more

### CODE REVIEW

File: cypress/tests/open-ebooks/server-e2e/login.ts [hunk 1]
```diff
@@ -151,3 +152,58 @@ describe("FirstBook login", () => {
...
+
+    // The redirect url is encoded twice, so we have to match that pattern in the test
+    const encodedRedirectURL = encodeURIComponent(
+      encodeURIComponent(Cypress.config().baseUrl as string)
+    );
+
+    cy.location().should(loc => {
+      expect(loc.hostname).to.eq("clever.com");
+      expect(loc.search).to.contain(encodedRedirectURL);
+    });
```
Comment (samanthaandrews) [inline cypress/tests/open-ebooks/server-e2e/login.ts:181]: This is the assertion where we are checking redirect to Clever. If we remove this, then we can also remove the `chromeWebSecurity:false` flag in `cypress.json`.

File: cypress.json [hunk 2]
```diff
@@ -1,4 +1,5 @@
 {
   "baseUrl": "http://localhost:3000",
-  "integrationFolder": "cypress/tests"
+  "integrationFolder": "cypress/tests",
+  "chromeWebSecurity": false
```
[the remark this answers is not in the document]
Comment (kristojorg) [reply cypress.json:4]: Should we just check that window.location was reset to something else (and thus a redirect will happen)? Or do we need to actually let the browser redirect?

[the remark this answers is not in the document]
Comment (kristojorg) [reply cypress.json:4]: It doesn't seem like we should need to disable this. There is a limited amount we can test with Clever login:
- That we redirect to the right URL at the right time
- That when clever sends the user back to us with a token in the url bar, we accept it and log the user in successfully. 
- That when clever sends the user back with an error, we handle that

Everything in between is Clever's responsibility and I don't think we need to test it

File: cypress/tests/open-ebooks/server-e2e/login.ts (hunk 1 above)
Comment (kristojorg) [reply cypress/tests/open-ebooks/server-e2e/login.ts:181]: Can we check that the browser _attempted_ to redirect to the right url?

**Review by kristojorg — COMMENTED** [review]
Some comments

File: cypress/tests/open-ebooks/server-e2e/login.ts (hunk 1 above)
Comment (samanthaandrews) [reply cypress/tests/open-ebooks/server-e2e/login.ts:181]: Unfortunately, I don't think so. I tried to do something like this, where we just count the number of url:change events that happen:

```
it('confirm redirect', () => {

  let urlRedirects = 0;
  cy
    .on('url:changed', () => {
      urlRedirects++;
    });

  cy.get('a').click('')

  cy
    .then(() => {
      expect(urlRedirects).to..eq(1);
    });
});
```
but the issue is that within Cypress we can't click on the button that redirects to a different superdomain. So I removed the code that requires us to disable chrome web security. It doesn't give us as much confidence that the redirect happened, but I think it's enough. What do you think?

File: cypress/tests/open-ebooks/integration/browse.ts [hunk 3]
```diff
@@ -161,34 +162,21 @@ describe("All-access browsing", () => {
...
-    cy.findByRole("link", { name: "See more: High School collection" }).click();
-    cy.wait("@highSchoolCollection");
-    cy.findByRole("link", { name: "See more: Staff Picks collection" }).click();
+    cy.log("navigate to a collection path");
+    cy.visit(HIGH_SCHOOL_STAFF_PICKS_COLLECTION_PATH);
     cy.wait("@staffPicks");
-    cy.findByRole("link", { name: "View Betrayed" }).click();
-    cy.wait("@betrayedBook");
-    // Verify breadcrumbs list exists
+    cy.wait(2000);
```
[the remark this answers is not in the document]
Comment (samanthaandrews) [reply cypress/tests/open-ebooks/integration/browse.ts]: Good catch. This test was flaky but I think I fixed it to use a .then() that makes it wait for the request before looking for the breadcrumbs list.

File: cypress/tests/open-ebooks/integration/login.ts [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+
+describe("FirstBook login", () => {
+  beforeEach(() => {
+    cy.clearCookie(COOKIE_KEY);
+  });
+
+  it("Can login with FirstBook", () => {
+    cy.intercept("GET", `${SERVER_URL}/groups`, {
+      fixture: "open-ebooks/all-access/all-collections.html"
+    }).as("allCollections");
```
[the remark this answers is not in the document]
Comment (samanthaandrews) [reply cypress/tests/open-ebooks/integration/login.ts]: Yeah it seems to be as long as a user is logged in with a good access token. I'm stubbing out this intercept before each test and then using `cy.wait("@allCollections)` in the test(s) where we visit the "/" home page.

Screen Shot 2021-09-06 at 9 47 56 PM

File: cypress/tests/open-ebooks/server-e2e/login.ts [hunk 5]
```diff
@@ -151,3 +152,46 @@ describe("FirstBook login", () => {
...
+      CLEVER_LOGIN_PATH
+    );
+
+    cy.findByRole("link", { name: "Login with Clever" }).should("exist");
+  });
+
+  it("Correct Clever token is handled", () => {
+    cy.visit(
+      `${APP_PATH}#access_token=${Cypress.env(
+        "OPENEBOOKS_CLEVER_HIGH_SCHOOL_ACCESS_TOKEN"
```
Comment (samanthaandrews) [inline cypress/tests/open-ebooks/server-e2e/login.ts:175]: This token was previously working but now it's not... (I grabbed it from manually logging in via Clever a few weeks ago). I'm guessing this means these tokens have an expiration date/time. I'm not exactly sure how to test that this is working. Any ideas?

(I am passing this token in as a variable in `.env.local`).

File: cypress/tests/open-ebooks/server-e2e/login.ts (hunk 1 above)
Comment (kristojorg) [reply cypress/tests/open-ebooks/server-e2e/login.ts]: Hmmm interesting, let's chat about this today

File: cypress/tests/open-ebooks/integration/login.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (kristojorg) [reply cypress/tests/open-ebooks/integration/login.ts]: This looks like the "/app" page to me though right?

[the remark this answers is not in the document]
Comment (samanthaandrews) [reply cypress/tests/open-ebooks/integration/login.ts:58]: Yes, my bad. It's definitely making the call only on calls to "/app" when a user is logged in.

[the remark this answers is not in the document]
Comment (kristojorg) [reply cypress/tests/open-ebooks/integration/login.ts:58]: Ok nice 👍

**Review by kristojorg — APPROVED** [review]
LGTM! Nice work : )

**Pull request merged**